### DISCUSSION TITLE
How can memory usage reached 20GB+ ?

### DISCUSSION THREAD

**Discussion by dikkoinfokes** [allinurl/goaccess#2734] [Q&A]
I'm curious why the GoAccess service consumes so much memory and continues to increase.

In my case, GoAccess uses an additional 1GB of memory each day, reaching up to 24.3GB. My server performs log rotation every week, and the log file size ranges between 6GB and 10GB each week. The memory is only released when the service is restarted. I assume this behavior is related to log data being cached.

If my assumption is correct, is there a way to control the memory usage? For example, could we implement a periodic flush or restart the service weekly?

This is goaccess process memory usage in my server:
![Screenshot 2024-10-10 092530](https://github.com/user-attachments/assets/9fbc85da-9db7-40e4-ad94-beed782500fc)

This is the command that I use:
`/usr/bin/goaccess -f /var/log/nginx/access.log --real-time-html --ws-url=wss://my-server-url:443/ws -o /var/www/html/analytics/report.html --port=7890 -p /etc/goaccess/goaccess.conf --origin=https://my-server-url --date-spec=now-2w`

This is the report that tell me the process read 108 GB data and use 24.3GB memory 
![image](https://github.com/user-attachments/assets/e6ab23f6-c45b-4bc7-a041-feee51363067)

**Comment by allinurl** [comment]
Thanks for sharing that! Just to clarify, are you looking to keep a week’s worth of data? You can use `--keep-last=<num_days>`, and GoAccess will manage the resources related to that data, e.g., `--keep-last=7`.

Can I ask how many requests you typically parse in a day, or for that 24GB log file, how many requests are you receiving? By the way, I’ve been working on a disk implementation, we will see how it evolves, so stay tuned for updates on that!

**Reply by dikkoinfokes** [reply]
Thank you for your response!

At the moment, my main concern is managing server memory usage. From what I’ve observed, if things continue as they are, my server might crash unless GoAccess is restarted. Regarding the log data, I can always trace older logs if needed, so keeping a week’s worth of data should be sufficient for monitoring purposes.

My log file size averages between 6~10GB per week, with around 1.5M transactions per day. The 24GB thing refers to the RAM used by the GoAccess service after running for over two months.

I’ll definitely keep an eye out for updates and will try your suggestion to limit the data to a week.

**Comment by allinurl** [comment]
`--date-spec=now-2w` doesn't limit retained data, `--date-spec` only controls date granularity. Use `--keep-last=14` (or the desired number of days) to bound storage. 

GoAccess `v1.11` also reduced storage memory usage by about `20%`. Closing this as answered, but feel free to reopen if needed.
